Mathlib/Topology/Germ.lean

```diff
@@ -0,0 +1,192 @@
+      intro f f' hff' hf
+      apply (hf.and <| Eventually.eventually_nhds hff').mono
+      rintro y ⟨hy, hy'⟩
+      rwa [Germ.coe_eq.mpr (EventuallyEq.symm hy')]
+    fun f f' hff' ↦ propext <| forall_congr' fun _ ↦ ⟨this f f' hff', this f' f hff'.symm⟩
+
+theorem Filter.Eventually.germ_congr_set
+    {P : ∀ x : X, Germ (𝓝 x) Y → Prop} (hf : ∀ᶠ x in 𝓝ˢ A, P x f)
+    (h : ∀ᶠ z in 𝓝ˢ A, g z = f z) : ∀ᶠ x in 𝓝ˢ A, P x g := by
+  rw [eventually_nhdsSet_iff_forall] at *
+  intro x hx
+  apply ((hf x hx).and (h x hx).eventually_nhds).mono
+  intro y hy
+  convert hy.1 using 1
+  exact Germ.coe_eq.mpr hy.2
+
+theorem restrictGermPredicate_congr {P : ∀ x : X, Germ (𝓝 x) Y → Prop}
+    (hf : RestrictGermPredicate P A x f) (h : ∀ᶠ z in 𝓝ˢ A, g z = f z) :
+    RestrictGermPredicate P A x g := by
+  intro hx
+  apply ((hf hx).and <| (eventually_nhdsSet_iff_forall.mp h x hx).eventually_nhds).mono
+  rintro y ⟨hy, h'y⟩
+  rwa [Germ.coe_eq.mpr h'y]
+
+theorem forall_restrictGermPredicate_iff {P : ∀ x : X, Germ (𝓝 x) Y → Prop} :
+    (∀ x, RestrictGermPredicate P A x f) ↔ ∀ᶠ x in 𝓝ˢ A, P x f := by
+  rw [eventually_nhdsSet_iff_forall]
+  rfl
+
+theorem forall_restrictGermPredicate_of_forall
+    {P : ∀ x : X, Germ (𝓝 x) Y → Prop} (h : ∀ x, P x f) :
+    ∀ x, RestrictGermPredicate P A x f :=
+  forall_restrictGermPredicate_iff.mpr (eventually_of_forall h)
+end RestrictGermPredicate
+
+namespace Filter.Germ
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `X → Z` at `x ∈ X` -/
+def sliceLeft [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.1) Z :=
+  P.compTendsto (Prod.mk · p.2) (Continuous.Prod.mk_left p.2).continuousAt
+
+@[simp]
+theorem sliceLeft_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceLeft = fun x' ↦ f (x', y) :=
+  rfl
+
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `Y → Z` at `y ∈ Y` -/
+def sliceRight [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.2) Z :=
+  P.compTendsto (Prod.mk p.1) (Continuous.Prod.mk p.1).continuousAt
+
+@[simp]
+theorem sliceRight_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceRight = fun y' ↦ f (x, y') :=
+  rfl
+
+end Filter.Germ
+
+/-- If the germ of `f` w.r.t. each `𝓝 x` is constant, `f` is locally constant. -/
+private lemma IsLocallyConstant.of_germ_isConstant (h : ∀ x : X, (f : Germ (𝓝 x) Y).IsConstant) :
+    IsLocallyConstant f := by
+  intro s
+  rw [isOpen_iff_mem_nhds]
+  intro a ha
+  obtain ⟨b, hb⟩ := h a
+  apply mem_of_superset hb
+  intro x hx
+  have : f x = f a := (mem_of_mem_nhds hb) ▸ hx
+  rw [mem_preimage, this]
+  exact ha
+
+-- should follow from `isConstant_compTendsto` specialised to the neigbourhood filter
+proof_wanted isConstant_comp_subtype {s : Set X} {f : X → Y} {x : s}
```

Comment:
This one is now `isConstant_comp_tendsto hf continuousAt_subtype_val`. Maybe move it a few lines up, in the namespace `Filter.Germ`?